operator/pkg/controller/virtualippool/controller.go

```diff
@@ -0,0 +1,425 @@
+package virtualippool
+
+import (
+	"context"
+	"crypto/sha256"
+	"encoding/hex"
+	"fmt"
+	"math"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/tools/record"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+
+	metalk8sscalitycomv1alpha1 "github.com/scality/metalk8s/operator/api/v1alpha1"
+	"github.com/scality/metalk8s/operator/pkg/controller/utils"
+)
+
+type VirtualIPPoolReconciler struct {
+	client   client.Client
+	scheme   *runtime.Scheme
+	recorder record.EventRecorder
+
+	instance metalk8sscalitycomv1alpha1.VirtualIPPool
+	// Some cache to ensure we don't reuse Virtual Router IDs, even for different pools
+	usedVRID       map[int]bool
+	nodeList       corev1.NodeList
+	configChecksum string
+}
+
+const (
+	// Name of the controller
+	controllerName = "virtualippool-controller"
+
+	// Name of the component
+	componentName = "metalk8s-vips"
+
+	// Name of the application
+	appName = "virtualippool"
+
+	// Annotation key to store Config checksum
+	annotationChecksumName = "checksum/config"
+
+	// ConfigMap key for HL config
+	hlConfigMapKey = "hl-config.yaml"
+)
+
+var log = logf.Log.WithName(controllerName)
+
+// Create a new VirtualIPPoolReconciler
+func newVirtualIPPoolReconciler(mgr ctrl.Manager) *VirtualIPPoolReconciler {
+	return &VirtualIPPoolReconciler{
+		client:   mgr.GetClient(),
+		scheme:   mgr.GetScheme(),
+		recorder: mgr.GetEventRecorderFor(controllerName),
+	}
+}
+
+// Add create the new Reconciler
+func Add(mgr ctrl.Manager) error {
+	reconciler := newVirtualIPPoolReconciler(mgr)
+
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&metalk8sscalitycomv1alpha1.VirtualIPPool{}).
+		Owns(&corev1.ConfigMap{}).
+		Owns(&appsv1.DaemonSet{}).
+		Complete(reconciler)
+}
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+func (r *VirtualIPPoolReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
```

Comment:
nit: I'm not a big fan of naming vars with such short name, reviewing it is not sufficiently explicit in my mind
```suggestion
func (virtualIPPoolReconciler *VirtualIPPoolReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
```

Reply:
I agree, but it's kind of "standard" in Go from what I saw

Reply:
Indeed, it seems that the short name indicates the object to which the current method applies. In storage-operator, we replaced it with `self`, similar to what we use in Python. Not sure which is best, but I think it's useful to differenciate this "method holder" from other variables.